File: src/main/java/com/warren/bt/Cooldown.java
package com.warren.bt;

import java.util.function.Supplier;

public class Cooldown extends Decorator {
  private final Supplier<Integer> cooldownDurationSupplier;
  private long lastExecutionTime = -1;

  public Cooldown(Node child, int cooldownDuration) {
    this(child, () -> cooldownDuration);
  }

  public Cooldown(Node child, Supplier<Integer> cooldownDurationSupplier) {
    super(child);
    this.cooldownDurationSupplier = cooldownDurationSupplier;
  }

  @Override
  public Status tick() {
    if (child == null) {
      return Status.FAILURE;
    }

    long currentTime = System.currentTimeMillis();
    int cooldownDuration = Math.max(0, cooldownDurationSupplier.get());

    if (lastExecutionTime >= 0) {
      long elapsed = currentTime - lastExecutionTime;
      if (elapsed < cooldownDuration) {
        return Status.FAILURE;
      }
    }

    Status status = child.tick();

    if (lastExecutionTime < 0 || status != Status.RUNNING) {
      lastExecutionTime = currentTime;
    }

    return status;
  }

  @Override
  public void reset() {
    super.reset();
    lastExecutionTime = -1;
  }
}

File: src/main/kotlin/com/warren/bt/Cooldown.kt
package com.warren.bt

/**
 * Simple cooldown: wait for [ticks] RUNNING frames before letting the child execute once. After the
 * child completes (success/failure), resets and starts cooldown again.
 */
class Cooldown(private val ticks: Int) : Decorator("Cooldown") {
  private var remaining = ticks
  private var runningChild = false

  override fun tick(): Status {
    val c = child ?: return Status.FAILURE

    if (!runningChild) {
      if (remaining > 0) {
        remaining--
        return Status.RUNNING
      }
      runningChild = true
    }

    val s = c.tick()
    if (s != Status.RUNNING) {
      // completed; start cooldown again
      c.reset()
      runningChild = false
      remaining = ticks
    }
    return s
  }

  override fun onReset() {
    remaining = ticks
    runningChild = false
  }
}
